src/cycleProcessing/batch.cpp

```diff
@@ -0,0 +1,227 @@
+ * dataProcessingConditions.requiredExtractedPointsCount will be added to batch
+ *
+ * @param [in] mediaInputStruct
+ * @param [in] dataProcessingConditions
+ * @param [in,out] currentBatch batch can contains some frames which was considered as bad for previous frame
+ *     but can be good for new one
+ *
+ * @return skipped frames count
+ */
+static int fillVideoFrameBatch(
+	MediaSources &mediaInputStruct,
+	const DataProcessingConditions &dataProcessingConditions,
+	std::vector<BatchElement> &currentBatch
+);
+
+/**
+ * Find new frame from batch which can be up to dataProcessingConditions.batchSize
+ *
+ * @param [in] mediaInputStruct
+ * @param [in] dataProcessingConditions
+ * @param [in,out] currentBatch <ul>
+ * 		<li>Input is a batch tail from previous iteration</li>
+ * 		<li>
+ * 			Output is a batch tail after new good frame
+ * 			(consequently there will be full batch in case we didn't find new good one)
+* 		</li>
+ * </ul>
+ * @param [in] previousFrame
+ * @param [out] newGoodFrame
+ * @param [in] previousFeatures
+ * @param [out] newFeatures
+ * @param [out] matches
+ * @return found frame batch index or some of error codes described above
+ */
+static int findGoodFrameFromBatch(
+	const DataProcessingConditions &dataProcessingConditions,
+	std::vector<BatchElement> &currentBatch,
+	std::vector<bool> &isMatchesEstimated,
+	std::vector<std::vector<DMatch>> &estimatedMatches,
+	Mat &previousFrame, Mat &newGoodFrame,
+	std::vector<KeyPoint> &previousFeatures,
+	std::vector<KeyPoint> &newFeatures,
+	std::vector<DMatch> &matches
+);
+
+int findGoodFrameFromBatchMultithreadingWrapper(
+	MediaSources &mediaInputStruct,
+	const DataProcessingConditions &dataProcessingConditions,
+	std::vector<BatchElement> &currentBatch,
+	Mat &previousFrame, Mat &newGoodFrame,
+	std::vector<KeyPoint> &previousFeatures,
+	std::vector<KeyPoint> &newFeatures,
+	std::vector<DMatch> &matches
+) {
+	fillVideoFrameBatch(mediaInputStruct, dataProcessingConditions, currentBatch);
+	int currentBatchSz = currentBatch.size();
+	if (currentBatchSz == 0)
+		return EMPTY_BATCH;
+
+	logStreams.mainReportStream << "Prev. extracted: " << previousFeatures.size() << std::endl;
+	ChronoTimer timer;
+
+	std::vector<bool> isMatchesEstimated(currentBatchSz, false);
+	int threadsCnt = dataProcessingConditions.threadsCount;
+	std::vector<std::thread> threads;
+	volatile bool threadsShouldDie = false;
+	std::vector<std::vector<DMatch>> estimatedMatches(currentBatchSz, std::vector<DMatch>());
+	for (int i = 0; i < threadsCnt; ++i) {
+		threads.emplace_back(std::thread([&](int threadIndex) {
+			for (
+				int batchIndex = currentBatchSz - 1 - threadIndex;
+				batchIndex >= 0;
+				batchIndex -= threadsCnt
+			) {
+				matchFramesPairFeatures(previousFrame, currentBatch.at(batchIndex).frame,
+					previousFeatures, currentBatch.at(batchIndex).features,
+					dataProcessingConditions.matcherType, estimatedMatches.at(batchIndex));
+				isMatchesEstimated.at(batchIndex) = true;

```

Comment:
Вчитывался несколько раз, но так и не понял, как из этого следует, что возвращение ссылки через `.at()` будет блокировать поток

Reply:
Вот что нашёл в стандарте C++11 (на самом деле цитату документации со стороннего сайта, но сейчас грузить ПДФку на 10 МБ меня не тянет):

23.2.2 Container data races [container.requirements.dataraces]

For purposes of avoiding data races (17.6.5.9), implementations shall
consider the following functions to be const: begin, end,
rbegin, rend, front, back, data, find, lower_bound,
upper_bound, equal_range, at and, except in associative or
unordered associative containers, operator[].

Notwithstanding (17.6.5.9), implementations are required to avoid data
races when the contents of the contained object in different elements
in the same sequence, excepting vector<bool>, are modified
concurrently.

Reply:
Вроде как тут нам говорят, что тут могут возникнуть рэйс кондишены - и это полностью наша ответственность

Reply:
Это ты где прочитал? Там же написано что для векторов это не относится

Reply:
Мне вообще не понятно, что написано в этих двух абзацах

Reply:
> Вчитывался несколько раз, но так и не понял, как из этого следует, что возвращение ссылки через `.at()` будет блокировать поток
Ну потому что если открыть реализацию matchFramesPairFeatures, то мы поймём, что дела очень плохи, потому что все потоки одновременно изменяют глобальную структуру данных.

Reply:
С какой мутабельной глобальной структурой мы работаем в этой функции?

Reply:
Всё потоки одновременно пишут в estimatedMatches: он подаётся в matchFramesPairFeatures, там в matchFeatures потом в getGoodMatches и там каждый поток в цикле пишет в свой вектор из вектора estimatedMatches. Кажется, в случае блокировки, ситуация очень плохая

Reply:
1. Очень сомневаюсь, что в векторе реализованы блокировки. Даже в Java потокобезопасные коллекции - отдельный класс, а в плюсах, где из встроенного только база, это сделали бы сразу? Звучит сомнительно. Но это не пруф, конечно. Надо точно понять. Однако пока что я склоняюсь к тому, что никакой потокобезопасности. Впрочем, это и не важно, потому что...
2. `.at()` возвращает нам указатель на элемент вектора - и далее мы работаем лишь с указателем на элемент - не с вектором как таковым

Reply:
Пойду для эксперимента делать страшные вещи...

Reply:
Узрите страшнейшего монстра из самых недр некромантии... DATA RACE!!! *Удар молнии*
```c++
void increment(int &num) {
	int tmp = num;
	this_thread::sleep_for(chrono::milliseconds(1));
	num = tmp + 1;
}

int main(int argc, char** argv) {
	std::vector<int> oneInt;
	std::vector<std::thread> threads;
	oneInt.push_back(0);
	for (int i = 0; i < 1000; ++i) {
		threads.push_back(std::thread([&]() {
			increment(oneInt.at(0));
		}));
	}
	for (auto &thread : threads)
		thread.join();
	std::cout << oneInt.at(0) << std::endl;
	return 0;
}
```
Вывод при разных запусках:
```sh
116
99
96
90
91
102
# ...
# Portal to the hell
```

Reply:
А если заставить поток засыпать на 5 секунд, то мы получаем ожидаемый результат - ни один поток не успеет увеличить значение до прочтения элемента каким-либо другим потоком, поэтому все запишут в `tmp` нуль, а потом присвоят единицу

Reply:
Эксперимент абсолютно бессмысленный, вообще не то надо было делать. Мы же хотим проверить, есть ли блокировка на векторе, а этот эксперимент вообще не отвечает на вопрос.

Reply:
Почему?

Reply:
Потому что если потоки будут последовательно изменять переменную, то есть если есть блокировка, то у нас будет тоже самое. 
1: записывает значение в тмп 
1:увеличивает тмп
1:пытается записать но блокируется 
2: записыает значение в тмп 
2: увеличивает тмп
2: записывает тмп в вектор
1:разблокируется и записывает тмп в вектор 
Итого в ответе значение из первого потока.

Reply:
То есть ты полагаешь, что блокировка может наступать именно в моменты доступа по ссылке?

Reply:
Ну да, именно когда мы пытаемся записать по ссылке

Reply:
Так лучше?
```c++
void slowIncrement(int &num) {
	num++;
}

int main(int argc, char** argv) {
	std::vector<int> oneInt;
	int singleNum = 0;
	volatile bool isAllStarted = false;
	std::vector<std::thread> threads;
	oneInt.push_back(0);
	for (int i = 0; i < 1000; ++i) {
		threads.push_back(std::thread([&]() {
			while (!isAllStarted)
				this_thread::yield();
			slowIncrement(oneInt.at(0));
			singleNum++;
		}));
	}
	isAllStarted = true;
	for (auto &thread : threads)
		thread.join();
	std::cout << oneInt.at(0) << std::endl;
	std::cout << singleNum << std::endl;
	return 0;
}
```
```sh
953
956

963
946

# ...
```

Reply:
Ну а вот вариант с близкой к нашей ситуации, то есть модификация вектора, являющегося элементом вектора:
```c++
void slowIncrement(std::vector<int> &nums) {
	nums.push_back(0);
}

int main(int argc, char** argv) {
	std::vector<std::vector<int>> oneVector(1);
	volatile bool isAllStarted = false;
	oneVector.at(0).clear();
	std::vector<std::thread> threads;
	for (int i = 0; i < 1000; ++i) {
		threads.push_back(std::thread([&]() {
			while (!isAllStarted)
				this_thread::yield();
			slowIncrement(oneVector.at(0));
		}));
	}
	isAllStarted = true;
	for (auto &thread : threads)
		thread.join();
	std::cout << oneVector.at(0).size() << std::endl;
	return 0;
}
```
Он будет падать с ошибками типа:
```sh
double free or corruption (fasttop)
double free or corruption (fasttop)
double free or corruption (fasttop)
double free or corruption (fasttop)
```

Reply:
Как мне кажется, вывод отсюда один - никакой защиты нет, поэтому добавление элемента и приводит к падению

Reply:
Хм, звучит логично

Reply:
Вот замеры:
```c++
void addNumsUpToN(std::vector<bool> &nums, int n) {
	for (int i = 0; i < n; ++i)
		nums.push_back(true);
}

int main(int argc, char** argv) {
	std::vector<std::vector<bool>> vectors(12, std::vector<bool>(0));
	std::vector<std::thread> threads;
	ChronoTimer vectorTimer;
	for (int i = 0; i < 12; ++i) {
		addNumsUpToN(vectors.at(i), 10000000);
//		threads.push_back(std::thread([&](int index) {
//			addNumsUpToN(vectors.at(index), 10000000);
//		}, i));
	}
	for (auto &thread : threads)
		thread.join();
//	vectorTimer.printLastPointDelta("12 threads time: ", std::cout); // 1305 1275 1343
	vectorTimer.printLastPointDelta("Single thread time: ", std::cout); // 3898 4327 3892
}
```

Reply:
Код с добавлением элементов и замерами:
```c++
void addNumsUpToN(std::vector<bool> &nums, int n) {
	for (int i = 0; i < n; ++i)
		nums.push_back(true);
}

int main(int argc, char** argv) {
	int threadsCnt = 12;
	std::vector<std::vector<bool>> vectors(threadsCnt, std::vector<bool>(0));
	std::vector<std::thread> threads;
	ChronoTimer vectorTimer;
	for (int i = 0; i < threadsCnt; ++i) {
		addNumsUpToN(vectors.at(i), 10000000);
//		threads.push_back(std::thread([&](int index) {
//			addNumsUpToN(vectors.at(index), 10000000);
//		}, i));
	}
	for (auto &thread : threads)
		thread.join();
//	vectorTimer.printLastPointDelta("12 threads time: ", std::cout); // 1305 1275 1343
	vectorTimer.printLastPointDelta("Single thread for 12 vectors time: ", std::cout); // 3898 4327 3892
	return 0;
}
```